src/main/java/io/confluent/copycat/hdfs/HdfsWriter.java

```diff
@@ -169,25 +294,45 @@ private boolean shouldRotate(TopicPartition topicPart) {
     return recordCounters.containsKey(topicPart) && recordCounters.get(topicPart) >= flushSize;
   }
 
-  private void rotate(TopicPartition topicPart)
-      throws CopycatException {
+  public void close() throws CopycatException {
+    List<Exception> exceptions = new ArrayList<>();
+    for (TopicPartition topicPart : assignment) {
+      try {
+        if (writers.containsKey(topicPart)) {
+          closeTempfile(topicPart);
+          appendToWAL(topicPart);
+          commitFile(topicPart);
+        }
+      } catch (IOException e) {
+        log.error("Error rotating {} when closing task.", tempFileNames.get(topicPart));
+        exceptions.add(e);
+      }
+    }
+    writers.clear();
+
+    for (WAL wal : wals.values()) {
+      try {
+        wal.close();
+      } catch (CopycatException e) {
+        log.error("Error closing {}.", wal.getLogFile());
+        exceptions.add(e);
+      }
+    }
+    wals.clear();
+
+    tempFileNames.clear();
+    recordCounters.clear();
+    offsets.clear();
+
     try {
-      RecordWriter<Long, SinkRecord> writer = writers.get(topicPart);
-      writer.close();
-      writers.remove(topicPart);
-      commitFile(topicPart);
+      storage.close();
     } catch (IOException e) {
-      throw new CopycatRuntimeException(e);
+      exceptions.add(e);
     }
-  }
 
-  private void commitFile(TopicPartition topicPart) throws IOException {
-    long offset;
-    offset = offsets.get(topicPart) + recordCounters.get(topicPart);
-    offsets.put(topicPart, offset);
-    String tempFileName = FileUtils.tempFileName(url, topicsDir, topicPart);
-    String finalFileName = FileUtils.committedFileName(url, topicsDir, topicPart, offset);
-    FileUtils.renameFile(fs, tempFileName, finalFileName);
+    if (exceptions.size() != 0) {
+      throw new CopycatException("Error closing files.");
```

Comment:
Maybe include at least the first exception as the cause? Or maybe include all of them as a list in the error message.

At a minimum, it probably makes sense to at least log them all -- at least the one during `storage.close()` is missing.